model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java

```diff
@@ -634,7 +657,14 @@ public KeycloakSession getKeycloakSession() {
         session.users().preRemove(realm, group);
 
         realm.removeDefaultGroup(group);
-        group.getSubGroupsStream().forEach(realm::removeGroup);
+
+        // TODO: this batch size could potentially be stored somewhere (config?) as a variable
+        // calculate out batches of subgroups to delete from the parent group to avoid grinding the server to a halt at large scale
+        // especially helpful for freeing up some amount of database time for other requests
+        long batches = (long) Math.ceil(session.groups().getSubGroupsCount(realm, group.getId()) / 1000.0);
```

Comment:
I didn't pick up on that feature of the entity manager but it totally makes sense, will resolve that. 

As far as the calculations go, I used the count as an indicator of where to stop but we could also just check if there are any groups left after each batch, it would result in more calls to the DB I think, though

Reply:
Let's see what the store team thinks about it. Perhaps there is a mechanism/pattern there already for batching.

What I was saying is to just include a provider configuration option with a default batch size and allow users to change it accordingly. We have this approach in other places.